cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -119,166 +124,461 @@ class RandomTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
+   override final def findNextNodeToVisit(): Option[Int] = {
+    if (queue.isEmpty) {
+      None
     } else {
-      val childId = nd.neighborIds(dir)(off)
-      if (onlyOnce && seenBefore(childId)) {
-        // try the next child
-        firstNotYetVisitedNode(nd, off + 1)
+      val next = queue.firstInt()
+      val visitedBefore = coloring.get(next) match {
+        case Walk.NodeColor.Visited => true
+        case _ => false
+      }
+      if (visitedBefore) {
+        handleVisitedInQueue(next)
+        findNextNodeToVisit()
       } else {
-        storeForVisit(nd.id, off)
-        Some(NodeDesc(childId, distance(nd.id).get + 1))
+        Some(next)
       }
     }
   }
 
-  def next = {
-    var NodeDesc(id, off) = stack.remove(stack.size - 1)
-    assert(off == -1)
-    val topnd = getExistingNodeById(graph, id)
-    val nextUnvisitedNode = firstNotYetVisitedNode(topnd, off + 1)
-    nextUnvisitedNode match {
-      case Some(NodeDesc(nodeId, dist)) => visitPotentialNode(nodeId, -1, dist)
-      case None => // do nothing
+  def handleVisitedInQueue(next: Int) {
+    queue.dequeueInt()
+  }
+
+  // can't limit number of nodes added to the queue, because we skip some nodes
+  override def limitAddedToQueue(nodes: Seq[Int]): Seq[Int] = nodes
+
+  override def shouldBeEnqueued(color: Walk.NodeColor.Color): Boolean = color != Walk.NodeColor.Visited
+}
+
+
+/**
+ * A queue based traverser trait that keeps track of visiting distance from `homeNodeIds`.
+ *
+ * Note that DepthTracker for the DFS case returns the depth of first seeing a particular
+ * node. It does not have to be the same as the depth at which the node is visited.
+ */
+trait PathLengthTracker extends QueueBasedTraverser {
+  private lazy val nextVisitDistanceTracker = new IntInfoKeeper(false)
+  private lazy val visitDistanceTracker = new IntInfoKeeper(true)
+
+  abstract override protected def enqueue(nodes: Seq[Int], from: Option[Int]): Unit = {
+    val fromDistance = from.flatMap(visitDistanceTracker.infoOfNode).getOrElse(-1)
+    nodes foreach { id =>
+      nextVisitDistanceTracker.recordInfo(id, fromDistance + 1)
     }
-    topnd
+    super.enqueue(nodes, from)
+  }
+
+  override def visitNode(node: Node) {
+    visitDistanceTracker.recordInfo(node.id, nextVisitDistanceTracker.infoOfNode(node).get)
+    super.visitNode(node)
   }
 
-  def hasNext = !stack.isEmpty
+  def distance(id: Int) = visitDistanceTracker.infoOfNode(id)
+}
+
+/**
+ * Trait to be mixed in to the DFS traverser to add finishing distance of a node.
+ */
+trait FinishingTimeTracker extends DepthFirstTraverser {
+  private var time = 0
+
+  private lazy val finishingTime = new IntInfoKeeper(true)
+
+  override def handleVisitedInQueue(next: Int): Unit = {
+    finishingTime.recordInfo(next, time)
+    super.handleVisitedInQueue(next)
+  }
+
+  override def visitNode(node: Node): Unit = {
+    time += 1
+    // adding again to capture finishing time of a node
+    queue.enqueueFirst(node.id)
```

Comment:
That may be a little wastefull, but:
1) We cannot deque in the findNextToDeque, because we need to have hasNext that does not dequeu. So we deque in the QueueBasedTraverser:next.
2) We can override QueueBasedTraverser:next in DepthFirstSearch or FinishingTimeTracker but wastefull adding when visiting looks more clear and concise for me.